controllers/front/OrderConfirmationController.php

```diff
@@ -45,31 +48,86 @@ public function init()
     {
         parent::init();
 
+        // If are coming to this page to finish free order, we need to run additional logic
         if (true === (bool) Tools::getValue('free_order')) {
             $this->checkFreeOrder();
         }
 
-        $this->id_cart = (int) (Tools::getValue('id_cart', 0));
-
-        $redirectLink = 'index.php?controller=history';
-
         $this->id_module = (int) (Tools::getValue('id_module', 0));
-        $this->id_order = Order::getIdByCartId((int) ($this->id_cart));
+        $this->id_order = (int) (Tools::getValue('id_order', 0));
         $this->secure_key = Tools::getValue('key', false);
-        $order = new Order((int) ($this->id_order));
+        $this->order = new Order((int) ($this->id_order));
 
+        // This data is kept only for backward compatibility purposes
+        $this->id_cart = (int) $this->order->reference;
+        $this->reference = (string) $this->order->reference;
+
+        // The confirmation link must contain a unique order secure key matching the key saved in database,
+        // this prevents user to view other customer's order confirmations
         if (!$this->id_order || !$this->id_module || !$this->secure_key || empty($this->secure_key)) {
             Tools::redirect($redirectLink . (Tools::isSubmit('slowvalidation') ? '&slowvalidation' : ''));
         }
-        $this->reference = $order->reference;
-        if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
+        $redirectLink = 'index.php?controller=history';
```

Comment:
```suggestion
        $redirectLink = $this->context->link->getPageLink('history', $this->ssl);
```
Avoid double redirection in case of Url Rewrite enabled